File: forge/templates/python-service-template/template/src/app/services/item_service.py
import logging
from uuid import UUID

from app.core.errors import AlreadyExistsError, NotFoundError
from app.data.repositories.item_repository import ItemRepository
from app.domain.item import (
    Item,
    ItemCreate,
    ItemStatus,
    ItemUpdate,
    PaginatedItemResponse,
)
from service.uow.aio import AsyncUnitOfWork

logger = logging.getLogger(__name__)


class ItemService:
    def __init__(self, uow: AsyncUnitOfWork):
        self._uow = uow

    async def list(
        self,
        *,
        skip: int = 0,
        limit: int = 50,
        status: ItemStatus | None = None,
        search: str | None = None,
    ) -> PaginatedItemResponse:
        async with self._uow as uow:
            repo = uow.repo(ItemRepository)
            items = await repo.list_items(skip=skip, limit=limit, status=status, search=search)
            total = await repo.count_items(status=status, search=search)

        return PaginatedItemResponse(
            items=list(items),
            total=total,
            skip=skip,
            limit=limit,
            has_more=(skip + limit) < total,
        )

    async def get(self, item_id: UUID) -> Item:
        async with self._uow as uow:
            repo = uow.repo(ItemRepository)
            item = await repo.get(item_id)
        if not item:
            raise NotFoundError("Item", item_id)
        return item

    async def create(self, data: ItemCreate) -> Item:
        async with self._uow as uow:
            repo = uow.repo(ItemRepository)
            if await repo.name_exists(data.name):
                raise AlreadyExistsError("Item", data.name)
            item = await repo.create(data)
        return item

    async def update(self, item_id: UUID, data: ItemUpdate) -> Item:
        async with self._uow as uow:
            repo = uow.repo(ItemRepository)
            existing = await repo.get(item_id)
            if not existing:
                raise NotFoundError("Item", item_id)
            if data.name and data.name != existing.name:
                if await repo.name_exists(data.name, exclude_id=item_id):
                    raise AlreadyExistsError("Item", data.name)
            item = await repo.update(item_id, data)
        return item

    async def delete(self, item_id: UUID) -> None:
        async with self._uow as uow:
            repo = uow.repo(ItemRepository)
            existing = await repo.get(item_id)
            if not existing:
                raise NotFoundError("Item", item_id)
            await repo.delete(item_id)

File: forge/templates/services/node-service-template/template/src/services/item.service.ts
import { AlreadyExistsError, NotFoundError } from "../lib/errors.js";
import {
	itemRepository,
	type PrismaItemRepository,
} from "../data/repositories/index.js";
import type { TenantContext } from "../middleware/tenant.js";
import type {
	ItemCreate,
	ItemStatus,
	ItemUpdate,
	PaginatedItems,
} from "../schemas/item.schema.js";

interface ListParams {
	tenant: TenantContext;
	skip: number;
	limit: number;
	status?: ItemStatus;
	search?: string;
}

/**
 * Service layer.
 *
 * Depends on :class:`PrismaItemRepository` only via the
 * :class:`Repository` interface — tests can substitute an in-memory
 * implementation without touching the database. The default export
 * uses the repository singleton; ``itemService.withRepository(repo)``
 * builds a service bound to a custom repo for tests.
 */

function buildService(repo: PrismaItemRepository = itemRepository) {
	return {
		async list(params: ListParams): Promise<PaginatedItems> {
			const { tenant, skip, limit, status, search } = params;
			const { items, total } = await repo.list(tenant, {
				skip,
				limit,
				status,
				search,
			});
			return {
				items,
				total,
				skip,
				limit,
				has_more: skip + items.length < total,
			};
		},

		async create(tenant: TenantContext, data: ItemCreate) {
			const existing = await repo.findByName(tenant, data.name);
			if (existing) throw new AlreadyExistsError("Item", data.name);
			return repo.create(tenant, data);
		},

		async getById(tenant: TenantContext, id: string) {
			const item = await repo.getById(tenant, id);
			if (!item) throw new NotFoundError("Item", id);
			return item;
		},

		async update(tenant: TenantContext, id: string, data: ItemUpdate) {
			await this.getById(tenant, id);
			if (data.name) {
				const existing = await repo.findByNameExcluding(tenant, data.name, id);
				if (existing) throw new AlreadyExistsError("Item", data.name);
			}
			return repo.update(tenant, id, data);
		},

		async remove(tenant: TenantContext, id: string) {
			await this.getById(tenant, id);
			await repo.delete(tenant, id);
		},

		withRepository(other: PrismaItemRepository) {
			return buildService(other);
		},
	};
}

const itemService = buildService();

export const list = itemService.list.bind(itemService);
export const create = itemService.create.bind(itemService);
export const getById = itemService.getById.bind(itemService);
export const update = itemService.update.bind(itemService);
export const remove = itemService.remove.bind(itemService);
export const withRepository = itemService.withRepository.bind(itemService);
